### PR TITLE
8256814: WeakProcessorPhases may be redundant

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change which eliminates the WeakProcessorPhase class.

The OopStorageSet class is changed to provide scoped enums for the different
categories: StrongId, WeakId, and Id (for the union of strong and weak).
An accessor is provided for obtaining the storage corresponding to a
category value.

Various other enumerator ranges, array sizes and indices, and iterations are
derived directly from the corresponding OopStorageSet category's enum range.

Iteration over a category of enumerators can be done via EnumIterator.  The
iteration over storage objects makes use of that enum iteration, rather than
having a bespoke implementation.  Some use-cases need iteration of the
enumerators, with storage lookup from the enumerator; other use-cases just
need the storage objects.

Testing:
mach5 tier1-6
Local (linux-x64) hotspot:tier1 with -XX:+UseShenandoahGC

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256814](https://bugs.openjdk.java.net/browse/JDK-8256814): WeakProcessorPhases may be redundant

### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 8d6c5903919deec96a7fad294fb149f2f9305fc3
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 3a4d5b78665ca473b259a9e2b3e7f02b6016e865
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to ebe50e35029b167dceb5c97a6c43f6e5b7c39de4

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1862/head:pull/1862`
`$ git checkout pull/1862`

### CODE REVIEW

**Comment by kimbarrett** [comment]
/summary Remove WeakProcessorPhase, adding scoped enum categories to OopStorageSet.

File: test/hotspot/gtest/gc/shared/test_oopStorageSet.cpp [hunk 1]
```diff
@@ -44,52 +45,54 @@ class OopStorageSetTest : public ::testing::Test {
     return count;
   }
 
-  template <uint count>
-  static void check_iterator(OopStorageSet::Iterator it,
```
Comment (stefank) [inline test/hotspot/gtest/gc/shared/test_oopStorageSet.cpp:48]: All the functions you changed are named `_iterator` and tested OopStorageSet::Iterator. Now the name is the same, but instead they test the Range facility. I think these functions should be renamed. Alternatively, we keep the tests for the OopStorageSet::Iterator and create a new set for the Range?

**Review by stefank — APPROVED** [review]
I think this looks good. I have a few comments that I would like to get addressed, but they are not blockers if you want to proceed with what you have.

File: src/hotspot/share/gc/shared/oopStorageSet.cpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/oopStorageSet.cpp:2]: It's 2021 by now, I suggest to update these before pushing.

**Review by tschatzl — APPROVED** [review]

**Review by rkennke — APPROVED** [review]
Changes look good to me! I also ran some tests with Shenandoah and they look good too! Thanks!

**Comment by kimbarrett** [comment]
Thanks @stefank , @tschatzl , @rkennke for reviews.

/integrate

**Pull request closed without merging**